src/NuGet.Services.Metadata.Catalog.Monitoring/Validation/Test/ValidationContext.cs

```diff
@@ -77,37 +86,57 @@ public class ValidationContext
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _commonLogger = logger.AsCommon();
 
             _databasetimestampMetadataResource = sourceRepositories.V2.GetResource<IPackageTimestampMetadataResource>();
             _databasePackageRegistrationMetadataResource = sourceRepositories.V2.GetResource<IPackageRegistrationMetadataResource>();
             _v3PackageRegistrationMetadataResource = sourceRepositories.V3.GetResource<IPackageRegistrationMetadataResource>();
-
-            var commonLogger = logger.AsCommon();
+            _httpSourceResource = sourceRepositories.V3.GetResource<HttpSourceResource>();
 
             _databaseIndex = new Lazy<Task<PackageRegistrationIndexMetadata>>(
-                () => _databasePackageRegistrationMetadataResource.GetIndexAsync(Package, commonLogger, CancellationToken));
+                () => _databasePackageRegistrationMetadataResource.GetIndexAsync(Package, _commonLogger, CancellationToken));
             _v3Index = new Lazy<Task<PackageRegistrationIndexMetadata>>(
-                () => _v3PackageRegistrationMetadataResource.GetIndexAsync(Package, commonLogger, CancellationToken));
+                () => _v3PackageRegistrationMetadataResource.GetIndexAsync(Package, _commonLogger, CancellationToken));
 
             _databaseLeaf = new Lazy<Task<PackageRegistrationLeafMetadata>>(
-                () => _databasePackageRegistrationMetadataResource.GetLeafAsync(Package, commonLogger, CancellationToken));
+                () => _databasePackageRegistrationMetadataResource.GetLeafAsync(Package, _commonLogger, CancellationToken));
             _v3Leaf = new Lazy<Task<PackageRegistrationLeafMetadata>>(
-                () => _v3PackageRegistrationMetadataResource.GetLeafAsync(Package, commonLogger, CancellationToken));
+                () => _v3PackageRegistrationMetadataResource.GetLeafAsync(Package, _commonLogger, CancellationToken));
 
             _timestampMetadataDatabase = new Lazy<Task<PackageTimestampMetadata>>(
                 () => _databasetimestampMetadataResource.GetAsync(this));
         }
 
+        public async Task<IEnumerable<IPackageSearchMetadata>> GetSearchPageForIdAsync(Uri baseUri)
+        {
+            return await _searchPageForIdCache.GetOrAdd(
+                baseUri,
+                _ =>
+                {
+                    var queryUri = new Uri(baseUri, "/query");
+                    _logger.LogInformation(
+                        "Searching for package {Id} on {Uri}.",
+                        Package.Id,
+                        queryUri.AbsoluteUri);
+
+                    var rawSearchResource = new RawSearchResourceV3(_httpSourceResource.HttpSource, new[] { queryUri });
+                    var packageSearchResource = new PackageSearchResourceV3(rawSearchResource);
+                    return packageSearchResource.SearchAsync(
+                        searchTerm: $"packageid:{Package.Id}",
```

Comment:
This may kick of the search request multiple times. Should this be wrapped by a `Lazy` to mitigate this?

Reply:
Good catch. TIL.